rfcs/20200113-tf-data-service.md

```diff
@@ -0,0 +1,591 @@
+be run from the beginning during master recovery.
+
+A concern with the above recovery strategy is that a master could transmit a
+split before crashing, then restart and transmit the same split again. To avoid
+this duplication, the master attaches a split index to every split it sends to a
+worker. When workers reconnect, they inform the master of their latest split
+index.
+
+Workers have no unrecoverable state. If a worker crashes, a new worker can take
+its place. It is up to the master to reassign splits from the crashed worker to
+the new worker.
+
+To improve worker recovery time, workers will periodically write checkpoints of
+their iterators to directories named using their worker ids. When the restarted
+worker connects, the master will tell it which iterator checkpoints to recover
+from.
+
+We will read and write this state through a MasterState interface which can be
+implemented using various storage backends. For use cases that require fault
+tolerance, the user must configure a fault-tolerant MasterState, e.g. Spanner
+internally, Cloud Spanner in GCP, or etcd externally. If fault tolerance isn't
+required, the user could configure state to be held in memory only.
+
+#### Leadership Transfer
+
+The master writes state to journal files so that the state can be recovered on
+restart. It is possible that a new master could be brought up while the old
+master is still running. If we aren't careful, this could result in corruption
+of the journal as both masters try to write to it.
+
+Ideally we could rely on a distributed coordination service such as ZooKeeper.
+However, this would add a significant burden to users who don't have access to a
+ZooKeeper cluster, and it would also require adding a new dependency on a
+ZooKeeper client.
+
+What TensorFlow does have is a FileSystem API. We will leverage this API to
+perform leadership transfer as follows:
+
+1.  The first master will create a file named "master_seqno_0". If it
+    successfully creates the file, it will consider itself the leader.
+1.  The leader master will check every N milliseconds that the "master_seqno"
+    file it created still exists. If the file no longer exists, the master will
+    cease operation immediately.
+1.  When a master thinks it should be leader, it attempts to atomically rename
+    the master_seqno_n file to master_seqno_n+1. If this succeeds, the master
+    will wait (N + M) milliseconds, verify that its renamed file still exists,
+    and begin acting as leader. This gives the previous leader time to notice
+    the rename.
+
+The above scheme relies on rename being atomic so that two masters don't both
+succeed at renaming the same file. Users may opt to use a filesystem that
+doesn't support atomic rename, but they do so at the (unlikely) risk of two
+concurrently running masters thinking they are leader. Common filesystems such
+as Posix and HDFS support atomic rename.
```

Comment:
And does GCS support it too?

Reply:
GCS doesn't support atomic rename. I think we could leverage GCS's strongly consistent directory listing to achieve something similar to atomic file rename, and loosen the requirement to "The filesystem supports either atomic rename or strongly consistent listing".